### PR TITLE
Update context tracking to use lifecycle callbacks

### PR DESCRIPTION
**Pull request by fractalwrench**
`ActivityManager#getRunningTasks` was deprecated on Lollipop and above. This PR changes the automatic Report context so that it is set as the name of the most recently launched `Activity` which is in the started state, or null if this condition cannot be met.

### CODE REVIEW

File: sdk/src/main/java/com/bugsnag/android/SessionTracker.java [hunk 1]
```diff
@@ -24,8 +25,8 @@
     private static final String KEY_LIFECYCLE_CALLBACK = "ActivityLifecycle";
     private static final int DEFAULT_TIMEOUT_MS = 30000;
 
-    private final ConcurrentHashMap<String, Boolean>
```
[the remark this answers is not in the document]
Comment (fractalwrench) [reply sdk/src/main/java/com/bugsnag/android/SessionTracker.java]: @kattrali I've changed it to a LinkedHashSet as that retains the ordering of values. The `testBasicInForeground` covers the case where multiple activities launch, then one stops, and we need to access the previous activity name.

Maybe a [ConcurrentLinkedDeque](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentLinkedDeque.html) would be a better choice for this, if you're happy with the general approach?

[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/main/java/com/bugsnag/android/SessionTracker.java]: Yeah the approach is solid - ConcurrentLinkedDeque seems ideal for this case as well.

[the remark this answers is not in the document]
Comment (fractalwrench) [reply sdk/src/main/java/com/bugsnag/android/SessionTracker.java]: Updated to use a ConcurrentLinkedDeque.

**Comment by kattrali** [comment]
Something went awry with the linter but I don't see what.

**Comment by fractalwrench** [comment]
Android Lint was failing as `ConcurrentLinkedDeque` is only available from API 21. I've changed the collection to use `ConcurrentLinkedQueue` instead, as that's been available from API 1, and we're not anticipating a large number of elements in this field.

**Pull request merged**